src/sqlfluff/linter.py

```diff
@@ -838,37 +839,68 @@ def lint_string(self, s, fname="<string input>", fix=False, config=None):
         return linted_file
 
     def paths_from_path(
-        self, path, ignore_file_name=".sqlfluffignore", ignore_non_existent_files=False
+        self,
+        path,
+        ignore_file_name=".sqlfluffignore",
+        ignore_non_existent_files=False,
+        ignore_files=True,
+        working_path=os.getcwd(),
     ):
         """Return a set of sql file paths from a potentially more ambigious path string.
 
         Here we also deal with the .sqlfluffignore file if present.
 
+        When a path to a file to be linted is explicitly passed
+        we look for ignore files in all directories that are parents of the file,
+        up to the current directory.
+
+        If the current directory is not a parent of the file we only
+        look for an ignore file in the direct parent of the file.
+
         """
         if not os.path.exists(path):
             if ignore_non_existent_files:
                 return []
             else:
                 raise IOError("Specified path does not exist")
 
-        # Files referred to exactly are never ignored.
-        if not os.path.isdir(path):
-            return [path]
+        # Files referred to exactly are also ignored if
+        # matched, but we warn the users when that happens
+        is_exact_file = not os.path.isdir(path)
+
+        if is_exact_file:
+            dirpath = os.path.dirname(path)
+            files = [os.path.basename(path)]
+            loader = ConfigLoader.get_global()
+            ignore_file_paths = loader.find_ignore_config_files(
+                path=path, working_path=working_path, ignore_file_name=ignore_file_name
+            )
+            path_walk_ignore_file = [
+                (
+                    os.path.dirname(ignore_file_path),
+                    None,
+                    [os.path.basename(ignore_file_path)],
+                )
+                for ignore_file_path in ignore_file_paths
+            ]
+            path_walk = [(dirpath, None, files)] + path_walk_ignore_file
```

Comment:
I *think* I can see what you're doing here, but this code is quite hard to read. I think it's down to the naming of `path_walk_ignore_file`, which maybe needs a clearer name.

If you can't think of a better name, could you add a few comments next to this code to explain how it works to people who might come across it in future?

Reply:
I added several comments, let me know what you think. Imo the naming is ok now that there are docs